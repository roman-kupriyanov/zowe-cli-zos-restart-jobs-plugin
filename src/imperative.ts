/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IImperativeConfig } from "@zowe/imperative";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    pluginHealthCheck: __dirname + "/healthCheck.Handler",
    pluginSummary: "Zowe CLI restart z/OS jobs plug-in",
    rootCommandDescription: "Plugin, which allows to restart z/OS jobs",
    productDisplayName: "Zowe CLI z/OS Jobs Restart Plug-in",
    name: "zos-restart-jobs"
};

export = config;
