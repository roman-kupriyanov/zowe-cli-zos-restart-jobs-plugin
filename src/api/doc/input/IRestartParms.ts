/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITaskWithStatus } from "@zowe/imperative";

/**
 * Interface for restart job API
 * @export
 * @interface IRestartParms
 */
export interface IRestartParms {

    /**
     * Returns spool content if this option used
     */
    viewAllSpoolContent?: boolean;

    /**
     * Wait for the job to reach output status
     */
    waitForActive?: boolean;


    /**
     * Wait for the job to reach output status
     */
    waitForOutput?: boolean;

    /**
     * Local directory path to download output of the job
     */
    directory?: string;

    /**
     * A file extension to save the job output with
     */
    extension?: string;

    /**
     * Task status object used by CLI handlers to create progress bars
     * for certain job restart requests
     * Optional
     */
    task?: ITaskWithStatus;

}
