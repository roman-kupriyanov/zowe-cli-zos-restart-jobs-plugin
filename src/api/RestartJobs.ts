/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IJob, GetJobs, SubmitJobs, JOB_STATUS } from "@zowe/cli";
import { AbstractSession, ImperativeError, ImperativeExpect } from "@zowe/imperative";
import { IRestartParms } from "./doc/input/IRestartParms";

/**
 * Class of restart jobs APIs for usage within the CLI and programmatically from node scripts
 * @export
 * @class RestartJobs
 */
export class RestartJobs {

    /**
     * Get JCL for job and modify it for a restart
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} stepname - name of a step, which to restart from
     * @param {IJob} job - job to get JCL for
     * @throws {ImperativeError} - throws an error if specified step name is not found in JCL
     * @returns {Promise<string>} - promise that resolves to a string, which contains JCL modified for a restart
     * @memberof RestartJobs
     */
    public static async getRestartJclForJob(session: AbstractSession, stepname: string, job: IJob) {
        const jobJcl: string = await GetJobs.getJclForJob(session, job);

        const newJclLines: string[] = [];
        const restartParamLine: string = `// RESTART=(${stepname.toUpperCase()})`;
        let isStepFound: boolean = false;
        let isRestartParmFound: boolean = false;

        for (const line of jobJcl.split("\n")) {

            // Transform to upper case to check for substrings
            const upperCasedLine = line.toUpperCase();

            // Line actually to be added in new JCL
            let lineToAdd: string = line.toString();

            // Do not try to process comment lines
            if (!upperCasedLine.startsWith("//*")) {

                if (upperCasedLine.indexOf("JOB") >= 0) {
                    // Remove redundant `jobid` at the end of JOB statement
                    lineToAdd = line.replace(job.jobid, "");
                }

                if (!isRestartParmFound) {
                    // Trim white spaces just in case to be able to check last line symbol
                    lineToAdd = lineToAdd.trim();

                    // If already specified RESTART= parm is found -> replace it with new step name
                    // and stop searching
                    if (lineToAdd.match(/RESTART=\(\S*\)/)) {
                        lineToAdd = lineToAdd.replace(/RESTART=\(\S*\)/, `RESTART=(${stepname})`);
                        isRestartParmFound = true;
                    }

                    // If no RESTART= parm is found inline and it is no continuation to next line ->
                    // add continuation and RESTART= parm on next line, then stop searching
                    if (!isRestartParmFound && !lineToAdd.endsWith(",")) {
                        lineToAdd += ",";
                        newJclLines.push(lineToAdd);
                        newJclLines.push(restartParamLine);
                        isRestartParmFound = true;
                        continue;
                    }
                }

                // Check if specified step name really exists in JCL
                if (upperCasedLine.indexOf("EXEC") >= 0) {
                    if (upperCasedLine.startsWith(`//${stepname.toUpperCase()}`)) {
                        isStepFound = true;
                    }
                }
            }

            newJclLines.push(lineToAdd);

        }

        if (!isStepFound) {
            throw new ImperativeError({
                msg: `Step name ${stepname} is not found in a job with jobid ${job.jobid}`
            });
        }

        return newJclLines.join("\n");

    }

    /**
     * Check if job is failed and return its JCL prepared for restart
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} jobid - job id to be translated into parms object
     * @param {string} stepname - a name of a step, which to restart from
     * @throws {ImperativeError} - throws an error if job is in ACTIVE state and return code is not "CC 0000"
     * @returns {Promise<IJob>} - Promise that resolves to an IJob document with details about the restarted job
     * @memberof RestartJobs
     */
    public static async getFailedJobRestartJcl(session: AbstractSession, jobid: string, stepname: string) {

        // Get the job details
        const job: IJob = await GetJobs.getJob(session, jobid);

        const errorMessagePrefix: string =
            `Restarting job with id ${jobid} on ${session.ISession.hostname}:${session.ISession.port} failed: `;

        ImperativeExpect.toBeEqual(job.status, JOB_STATUS.OUTPUT,
                                      errorMessagePrefix + "Job status is ACTIVE, OUTPUT is required");
        ImperativeExpect.toNotBeEqual(job.retcode, "CC 0000",
                                      errorMessagePrefix + "Job status is successful, failed is required");

        // Get the restart job JCL
        return this.getRestartJclForJob(session, stepname, job);

    }

    /**
     * Restart a job from a specific step
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} jobid - job id to be translated into parms object
     * @param {string} stepname - name of a step, which to restart from
     * @throws {ImperativeError} - throws an error if job is in ACTIVE state and return code is not "CC 0000"
     * @returns {Promise<IJob>} - promise that resolves to an IJob document with details about the restarted job
     * @memberof RestartJobs
     */
    public static async restartFailedJob(session: AbstractSession, jobid: string, stepname: string) {

        // Get the restart job JCL
        const restartJobJcl: string = await this.getFailedJobRestartJcl(session, jobid, stepname);

        // Re-submit restart job JCL
        return SubmitJobs.submitJcl(session, restartJobJcl);

    }

    /**
     * Restart a job from a specific step
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} jobid - job id to be translated into parms object
     * @param {string} stepname - name of a step, which to restart from
     * @param {IRestartParms} parms - special object with restart parameters (see for details)
     * @throws {ImperativeError} - throws an error if job is in ACTIVE state and return code is not "CC 0000"
     * @returns {Promise<IJob | ISpoolFile[]>} - promise that resolves to an IJob document with details about the restarted job or
     *                                           into a list of ISpoolFile documents with spool data set content
     * @memberof RestartJobs
     */
    public static async restartFailedJobWithParms(session: AbstractSession, jobid: string, stepname: string,
                                                  parms: IRestartParms) {

        // Get the restart job JCL
        const restartJobJcl: string = await this.getFailedJobRestartJcl(session, jobid, stepname);

        // Transform into ISubmitParms structure
        const submitParms = {
            jclSource: undefined as any,
            viewAllSpoolContent: parms.viewAllSpoolContent,
            directory: parms.directory,
            extension: parms.extension,
            waitForActive: parms.waitForActive,
            waitForOutput: parms.waitForOutput,
            task: parms.task
        };

        // Re-submit restart job JCL
        return SubmitJobs.submitJclString(session, restartJobJcl, submitParms);

    }

}
