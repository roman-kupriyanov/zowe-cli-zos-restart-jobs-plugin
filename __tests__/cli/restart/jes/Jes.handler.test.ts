/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ZosmfSession, IJob } from "@zowe/cli";
import { CommandProfiles, IHandlerParameters, IProfile, Session, TaskProgress, TaskStage } from "@zowe/imperative";
import { JesDefinition } from "../../../../src/cli/restart/jes/Jes.definition";
import JesHandler from "../../../../src/cli/restart/jes/Jes.handler";
import { RestartJobs } from "../../../../src/api/RestartJobs";
import { IRestartParms } from "../../../../src/api/doc/input/IRestartParms";

const host = "127.0.0.1";
const port = "443";
const user = "user";
const password = "password";
const protocol = "https";
const rejectUnauthorized = false;

const PROFILE_MAP = new Map<string, IProfile[]>();
PROFILE_MAP.set(
    "zosmf", [{
        name: "zosmf",
        type: "zosmf",
        host,
        port,
        user,
        password
    }]
);
const PROFILES: CommandProfiles = new CommandProfiles(PROFILE_MAP);
const DEFAULT_PARAMETERS: IHandlerParameters = {
    arguments: {$0: "", _: []}, // Please provide arguments later on
    response: {
        data: {
            setMessage: jest.fn((setMsgArgs) => {
                expect(setMsgArgs).toMatchSnapshot();
            }) as any,
            setObj: jest.fn((setObjArgs) => {
                expect(setObjArgs).toMatchSnapshot();
            }),
            setExitCode: jest.fn()
        },
        console: {
            log: jest.fn((logs) => {
                expect(logs.toString()).toMatchSnapshot();
            }) as any,
            error: jest.fn((errors) => {
                expect(errors.toString()).toMatchSnapshot();
            }) as any,
            errorHeader: jest.fn(() => undefined)
        },
        progress: {
            startBar: jest.fn((parms) => {
                expect(parms).toMatchSnapshot();
            }),
            endBar: jest.fn(() => undefined)
        },
        format: {
            output: jest.fn((parms) => {
                expect(parms).toMatchSnapshot();
            })
        }
    },
    definition: JesDefinition,
    fullDefinition: JesDefinition,
    profiles: PROFILES
};

describe("JesHandler tests", () => {

    const jobid: string = "JOB04541";
    const stepname: string = "STEP02";

    const defaultReturn: IJob = {
        owner: "",
        phase: 128,
        subsystem: "JES2",
        "phase-name": "Job is active in input processing",
        "job-correlator": "J0004565SVSCJES2D74D1C38.......:",
        type: "JOB",
        url: "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A",
        jobid,
        class: "A",
        "files-url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A/files",
        jobname: "TESTJOB",
        status: "INPUT",
        retcode: null
    };

    const defaultParms: IRestartParms = {
        viewAllSpoolContent: false,
        directory: "",
        extension: ".txt",
        waitForActive: false,
        waitForOutput: false,
        task: {
            statusMessage: "Restarting job",
            percentComplete: TaskProgress.TEN_PERCENT,
            stageName: TaskStage.IN_PROGRESS
        }
    };

    const testProfile = PROFILE_MAP.get("zosmf")[0];
    const testSession: Session = new Session({
        type: "basic",
        hostname: testProfile.host,
        port: testProfile.port,
        user: testProfile.user,
        password: testProfile.password,
        rejectUnauthorized,
        protocol
    });

    const createBasicZosmfSessionFromArgumentsSpy = jest.spyOn(ZosmfSession, "createBasicZosmfSessionFromArguments");
    const restartFailedJobWithParmsSpy = jest.spyOn(RestartJobs, "restartFailedJobWithParms");

    beforeEach(() => {

        createBasicZosmfSessionFromArgumentsSpy.mockClear();
        createBasicZosmfSessionFromArgumentsSpy.mockImplementation(() => testSession);

        restartFailedJobWithParmsSpy.mockClear();
        restartFailedJobWithParmsSpy.mockImplementation(() => Promise.resolve(defaultReturn));

    });

    it("should call the restartFailedJobWithParms API", async () => {

        const handler = new JesHandler();

        const commandParameters = {...DEFAULT_PARAMETERS};
        commandParameters.arguments = {
            ...commandParameters.arguments,
            jobid,
            stepname,
            host,
            port,
            user,
            password,
            rejectUnauthorized,
            protocol,
            viewAllSpoolContent: defaultParms.viewAllSpoolContent,
            directory: defaultParms.directory,
            extension: defaultParms.extension,
            waitForActive: defaultParms.waitForActive,
            waitForOutput: defaultParms.waitForOutput,
        };

        await handler.process(commandParameters);

        expect(createBasicZosmfSessionFromArgumentsSpy).toHaveBeenCalledTimes(1);
        expect(createBasicZosmfSessionFromArgumentsSpy).toHaveBeenCalledWith(commandParameters.arguments);

        expect(restartFailedJobWithParmsSpy).toHaveBeenCalledTimes(1);
        expect(restartFailedJobWithParmsSpy).toHaveBeenCalledWith(
            testSession,
            jobid,
            stepname,
            {
                viewAllSpoolContent: defaultParms.viewAllSpoolContent,
                directory: defaultParms.directory,
                extension: defaultParms.extension,
                waitForActive: defaultParms.waitForActive,
                waitForOutput: defaultParms.waitForOutput,
                task: defaultParms.task
            } as IRestartParms
        );

    });

});
