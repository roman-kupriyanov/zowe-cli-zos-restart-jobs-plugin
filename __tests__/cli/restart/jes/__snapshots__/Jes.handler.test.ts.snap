// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JesHandler tests should call the restartFailedJobWithParms API 1`] = `
Object {
  "task": Object {
    "percentComplete": 10,
    "stageName": 0,
    "statusMessage": "Restarting job",
  },
}
`;

exports[`JesHandler tests should call the restartFailedJobWithParms API 2`] = `
Object {
  "fields": Array [
    "jobid",
    "retcode",
    "jobname",
    "status",
  ],
  "format": "object",
  "output": Object {
    "class": "A",
    "files-url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A/files",
    "job-correlator": "J0004565SVSCJES2D74D1C38.......:",
    "jobid": "JOB04541",
    "jobname": "TESTJOB",
    "owner": "",
    "phase": 128,
    "phase-name": "Job is active in input processing",
    "retcode": null,
    "status": "INPUT",
    "subsystem": "JES2",
    "type": "JOB",
    "url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A",
  },
}
`;

exports[`JesHandler tests should call the restartFailedJobWithParms API 3`] = `
Object {
  "class": "A",
  "files-url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A/files",
  "job-correlator": "J0004565SVSCJES2D74D1C38.......:",
  "jobid": "JOB04541",
  "jobname": "TESTJOB",
  "owner": "",
  "phase": 128,
  "phase-name": "Job is active in input processing",
  "retcode": null,
  "status": "INPUT",
  "subsystem": "JES2",
  "type": "JOB",
  "url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A",
}
`;

exports[`JesHandler tests should call the restartFailedJobWithParms API 4`] = `"Successfully downloaded output to .//JOB04541"`;

exports[`JesHandler tests should call the restartFailedJobWithParms API 5`] = `"Restarted JCL with jobid \\"JOB04541\\" starting from step \\"STEP02\\""`;
